gpu-alloc/src/block.rs

```diff
@@ -150,29 +150,28 @@ impl<M> MemoryBlock<M> {
         size: usize,
     ) -> Result<NonNull<u8>, MapError> {
         let size_u64 = u64::try_from(size).expect("`size` doesn't fit device address space");
-        let size = align_up(size_u64, self.map_mask)
-            .expect("aligned `size` doesn't fit device address space");
-
-        let aligned_offset = align_down(offset, self.map_mask);
-
         assert!(offset < self.size, "`offset` is out of memory block bounds");
         assert!(
             size_u64 <= self.size - offset,
             "`offset + size` is out of memory block bounds"
         );
 
+        if !self.acquire_mapping() {
```

Comment:
This should be done after all asserts.

Reply:
You mean these two lines?
```rust
let offset_align_shift = isize::try_from(offset - aligned_offset)
                    .expect("`non_coherent_atom_size` is too large");
                let end = align_up(offset + size_u64, self.map_mask)
                    .expect("mapping end doesn't fit device address space");
```
Does it matter that the error is checked after those?

Reply:
If one of those panics, flag will be left in MAPPED state while memory is unmapped in fact.

Reply:
Are there any guarantees about the state of the program after panic? My understanding is that panics are called like that because they aren't handled properly, otherwise they would be errors.

Reply:
Panics represent situation that should not occur.

Typically API is designed in a way that if user can predict exactly that particular input will always cause an error then panic is used. One common example is Index trait.

But see how fetching invalid index never break type's invariants.

Same logic I apply in my APIs. Panic never causes internal inconsistency. I usually achieve this by placing panics before any mutations.

Reply:
I'm not sure if this approach is worth holding. You are basically suggesting to treat panics as error conditions that could be handled, and the data related to the panic to be potentially used again (hence the need for internal consistency), even though you are saying that "Panics represent situation that should not occur". A panic caught in gpu-alloc should mean that nobody should touch those gpu-alloc structures ever, and things are broken. Otherwise, it shouldn't be a panic, it should be a regular error.

Reply:
The cost of trying to follow this "good practice" is just too high. It's not visible to the developer where the panic can occur, so it's hard to preserve the invariants of the type. We have errors for this specifically.

Reply:
> A panic caught in gpu-alloc should mean that nobody should touch those gpu-alloc structures ever, and things are broken.

I don't know where you got this idea. Panic doesn't mean things are broken, panic means that API was used incorrectly. Regular errors are for problems occurred during correct API usage.

Reply:
I think you are contradicting yourself a little bit here, but this isn't worth fighting about.
If you want the panic-consistency, you should have the clippy check for all panics in the code and annotate them accordingly, at the very least.
The PR is now updated, please have another look!